### DISCUSSION TITLE
Would anyone be interested in being able to also select the column in a table, instead of just a row?

### DISCUSSION THREAD

**Discussion by hasezoey** [ratatui/ratatui#376] [Ideas]
Currently in base `widgets::Table`, you can only select a full row, i would like to know the interest in also being able to select the column in base ratatui, something along the lines of:

```rust
Table::new(vec![
    Row::new(vec!["Row11", "Row12", "Row13"]),
    Row::new(vec!["Row21", "Row22", "Row23"]).height(2),
    Row::new(vec!["Row31", "Row32", "Row33"]),
    Row::new(vec!["Row41", "Row42", "Row43"]).height(2),
])
.header(Row::new(vec!["Head1", "Head2", "Head3"]).bottom_margin(1))
.block(Block::default().borders(Borders::ALL))
.highlight_symbol(">> ")
.highlight_symbol_columns("> ") // maybe inherit from "highlight_symbol" if not explicitly provided?
.widths(&[
    Constraint::Length(5),
    Constraint::Length(5),
    Constraint::Length(5),
])
.column_spacing(1);

state.select(Some(0));
Buffer::with_lines(vec![
    "┌────────────────────────────┐",
    "│   Head1 Head2 Head3        │",
    "│                            │",
    "│>> Row11 Row12 Row13        │",
    "│   Row21 Row22 Row23        │",
    "│                            │",
    "│   Row31 Row32 Row33        │",
    "└────────────────────────────┘",
]);

state.select_with_column(Some((0, 1)));
Buffer::with_lines(vec![
    "┌────────────────────────────┐",
    "│   Head1   Head2 Head3      │",
    "│                            │",
    "│>> Row11 > Row12 Row13      │",
    "│   Row21   Row22 Row23      │",
    "│                            │",
    "│   Row31   Row32 Row33      │",
    "└────────────────────────────┘",
]);
```

and maybe in addition also something like `highlight_style_column(Style)`, for example: so that the selected column may be a different background color to better hint at which columns is selected.

in addition, how to handle the 0 column selection, should it just be the row selection (no column selection symbol), or should both be drawn?
OR
should the row symbol not be drawn at all if a column is selected (and it is up to the user to highlight the row via something like background color)

also as for some implementation detail, i think the `column_spacing` should be merged if a "selected column symbol allocation" has to be done (if the spacing is less than the symbol, expand to the length of the symbol with a space before the symbol, if it is more space then the symbol, deduct from the spacing the length of the symbol)

PS: i am not sure on how a "missing cell in a row" should be handled (example first row has 3 cells, second has 2, third may have 3)

**Comment by Valentin271** [comment]
First thing that came to my mind was [gobang](https://github.com/TaKO8Ki/gobang), a database browser where you can navigate **each cell**. So yes, I think this could be useful.
As for myself, I don't need it at the moment but could definitely use it later.

If ratatui doesn't provide an API way of selecting a cell, I think it should provide some workaround or way to implement it while not starting from scratch (somehow).
Right now, if someone needs it, he'll need to basically copy the table code and the the column selection, which is not a good thing in my opinion.

For the selection, I think an enum could be easier (for the maintainers and the users) than adding APIs to the table, something like

```rust
enum TableSelected {
    Row(u32),
    Col(u32, u32)
}
``` 

Maybe we should also look at what's been done so far, to see if someone had to implement it or could benefit from it.

**Reply by hasezoey** [reply]
> For the selection, I think an enum could be easier (for the maintainers and the users) than adding APIs to the table, something like

true, but for now i wanted to avoid breaking changes (ie let `select` stay the way it works currently), will have to know if the maintainers are interested in breaking this or not

**Reply by tonogdlp** [reply]
Yes!!! For me it is obvious that a table should be able to select a cell. Adding to Valentin271's comment, I think you could do a enum with column selection too:

```rust
enum TableSelected {
    Row(u32),
    Col(u32),
    Cell(u32, u32)
}
```

I think this functionality is a basic requirement of the crate!

**Comment by tonogdlp** [comment]
I took a look on how to make it possible for a Table to select a single cell. Here is the resulting "examples/table.rs" table..  I added some KeyCode handling to be able to toggle the new functionality. 


https://github.com/ratatui-org/ratatui/assets/8254361/002ff343-28ea-47ea-8778-1535ba87d4c4

Basically I made 3 changes, all in widgets/table.rs

1. Added a `struct` to represent a selection in the table 

```rust
pub struct TableSelection {
    pub row: usize,
    pub col: usize,
}
```

2. Changed `TableState` to use the new struct instead of `selected: Option<usize>`:
```rust
pub struct TableState {
    offset: usize,
    selected: Option<TableSelection>,
}
```


3. Added a couple of options to `Table` to be able to be somewhat backwards compatible. Specifically, I added a toggle for "full row highlighting", and another one to specify if the space for the selection (i.e, ">> ") should be added to only the selected column, or to all columns:
```rust 
pub struct Table<'a> {
    .......
    /// Wheter to highlight the entire row
    highlight_full_row: bool, 
    /// Whether to allocate spacing for all columns. Ignored if `HighlighSpacing == Never`
    highlight_spacing_all_columns: bool,
}
```

After this 3 changes, I updated the implementation of the table widget to use this new fields.

I will be doing a PR as soon as I have time to polish my code.. in the meantime I made a fork of this feature so you can check it out if you want to!

**Comment by Valentin271** [comment]
Looks great!

Could your PR include a column selection?

I think it would be quite easy to do something like so, just have to figure out the drawing then.
```rust
pub struct Table<'a> {
    .......
    /// Whether to highlight the entire row
    highlight_full_row: bool,

    /// Whether to highlight the entire col.
    /// Might be better to replaces those `bool`s with a bitflag enum (see Borders)
    highlight_full_col: bool,

    /// Whether to allocate spacing for all columns. Ignored if `HighlighSpacing == Never`
    highlight_spacing_all_columns: bool,
}
```

Like so we could select an entire row and col at the same time.

**Reply by tonogdlp** [reply]
Hey @Valentin271 !
I made a PR (#478) that includes the ability to highlight cells, rows, columns and both rows and columns. I went with an enum instead of a bitflag because a bitflag should be able to be summed and each sum should yield a specific, unique result. But I hadn't considered the ability for column selection, and I thank you for that!
